src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -400,4 +400,60 @@ object messages {
            |  ((${nestedRepresentation}))""".stripMargin
     }
   }
+
+  case class UnboundPlaceholderParameter()(implicit ctx:Context)
+    extends Message(14) {
+    val kind = "Syntax"
+
+    val msg = hl"unbound placeholder parameter; incorrect use of `_`"
+
+    val codeUnboundInBlock =
+      """
+        |{ _ }
+      """.stripMargin
+
+    val codeBoundInBlock =
+      """
+        |x => { x }
+      """.stripMargin
+
+    val codeUnboundVal =
+      """
+        |val a = _
+      """.stripMargin
+
+    val codeVar =
+      """
+        |var a = _
+      """.stripMargin
+
+    val explanation =
+      hl"""The `_` placeholder syntax was used where it could not be bound.
+        |Consider explicitly writing the variable binding.
+        |
+        |This can be done by replacing `_` with a variable (eg. `x`)
+        |and adding ${"x =>"} where applicable.
+        |
+        |Example before:
+        |
+        |$codeUnboundInBlock
+        |
+        |Example after:
+        |
+        |$codeBoundInBlock
+        |
+        |Another common occurrence for this error is defining a val with `_`:
+        |
+        |$codeUnboundVal
+        |
+        |But this val definition isn't very useful, it can never be assigned
+        |another value. And thus will always remain uninitialized.
+        |Consider replacing the ${"val"} with ${"var"}:
+        |
+        |$codeVar
+        |
+        |Note that this use of `_` is not placeholder syntax,
+        |but an uninitialized var definition
+        """.stripMargin
+  }
```

Comment:
Great job generalizing this error!

For the `explanation` I would inline the values `codeUnboundInBlock`, `codeBoundInBlock`, `codeUnboundVal` and `codeVar`, which would get the error message down from 55 lines to roughly 30.